www/js/lib/popovers.js

```diff
@@ -0,0 +1,455 @@
+/**
+ * popovers.js : Functions to add popovers to the UI
+ *
+ * Copyright 2013-2024 Jaifroid and contributors
+ * Licence GPL v3:
+ *
+ * This file is part of Kiwix.
+ *
+ * Kiwix is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public Licence as published by
+ * the Free Software Foundation, either version 3 of the Licence, or
+ * (at your option) any later version.
+ *
+ * Kiwix is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public Licence for more details.
+ *
+ * You should have received a copy of the GNU General Public Licence
+ * along with Kiwix (file LICENSE-GPLv3.txt).  If not, see <http://www.gnu.org/licenses/>
+ */
+
+'use strict';
+
+/* global params */
+
+import uiUtil from './uiUtil.js';
+
+/**
+ * Parses a linked article in a loaded document in order to extract the first main paragraph (the 'lede') and first
+ * main image (if any). This function currently only parses Wikimedia articles. It returns an HTML string, formatted
+ * for display in a popover
+ * @param {String} href The href of the article link from which to extract the lede
+ * @param {String} baseUrl The base URL of the currently loaded article
+ * @param {Document} articleDocument The DOM of the currently loaded article
+ * @param {ZIMArchive} archive The archive from which to extract the lede
+ * @returns {Promise<String>} A Promise for the linked article's lede HTML including first main image URL if any
+ */
+function getArticleLede (href, baseUrl, articleDocument, archive) {
+    const uriComponent = uiUtil.removeUrlParameters(href);
+    const zimURL = uiUtil.deriveZimUrlFromRelativeUrl(uriComponent, baseUrl);
+    console.debug('Previewing ' + zimURL);
+    // Do a binary search in the URL index to get the directory entry for the requested article
+    return archive.getDirEntryByPath(zimURL).then(function (dirEntry) {
+        const readArticle = function (dirEntry) {
+            // Wrap legacy callback-based code in a Promise
+            return new Promise((resolve, reject) => {
```

Comment:
You could probably do the same thing here:

```
function readArticleInner(resolve, reject) {
  ...
}

return new Promise(readArticleInner)
```

I feel like strategies like this help a lot when dealing with interwoven promise and callback code.